### PR TITLE
Oxc Transformer Alpha

### LINKED ISSUE #201: Announce oxc-transform alpha for react and ts transform

### PR DESCRIPTION
**Pull request by Boshen**
closes #201

### CODE REVIEW

File: src/blog/2024-09-29-transformer-alpha.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+```javascript
+let xform = oxc.transform(entry, source, {
+  typescript: {
+    onlyRemoveTypeImports: true,
+    declaration: {
+      stripInternal: true,
+    },
+  },
+});
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:43]:
```suggestion
const transformed = oxc.transform(entry, source, {
  typescript: {
    onlyRemoveTypeImports: true,
    declaration: { stripInternal: true },
  },
});
const {
  code, // Transformed to JS
  declaration, // TS declaration `.d.ts`
} = transformed;
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+The [Rolldown](https://rolldown.rs) bundler uses the `oxc_transformer` crate [directly](https://github.com/rolldown/rolldown/blob/3213e8fdb9e25f29295b6ec0d92fcc2ce03ce396/crates/rolldown/src/utils/pre_process_ecma_ast.rs#L67-L75).
+
+## [Benchmark Results](https://github.com/oxc-project/bench-transformer)
+
+On [GitHub Actions ubuntu-latest](https://github.com/oxc-project/bench-transformer/actions/workflows/ci.yml):
+
+Using the following files:
+
+- [TypeScript - parser.ts](https://github.com/microsoft/TypeScript/blob/3ad0f752482f5e846dc35a69572ccb43311826c0/src/compiler/parser.ts) - an atypical large file with 10777 lines.
+- [vuejs -renderer.ts](https://github.com/vuejs/core/blob/cb34b28a4a9bf868be4785b001c526163eda342e/packages/runtime-core/src/renderer.ts) - a large library file with 2550 lines.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:68]: Nit. Add a space to match the rest.

```suggestion
- [vuejs - renderer.ts](https://github.com/vuejs/core/blob/cb34b28a4a9bf868be4785b001c526163eda342e/packages/runtime-core/src/renderer.ts) - a large library file with 2550 lines.
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+Oxc is x faster than swc, and x faster than Babel.
+
+## Package size
+
+For package download size, oxc downloads two packages for a total of 2MB.
+
+| Package                                                                                  | Size                                                                                        |
+| ---------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------- |
+| `@oxc-transform/binding-darwin-arm64`                                                    | [1.95 MB](https://www.npmjs.com/package/@oxc-transform/binding-darwin-arm64)                |
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:110]: Nit. Leave off 1 decimal place so it looks visually smaller.

```suggestion
| `@oxc-transform/binding-darwin-arm64`                                                    | [1.9 MB](https://www.npmjs.com/package/@oxc-transform/binding-darwin-arm64)                |
```

**Comment by Dunqing** [comment]
Do we need to add an isolatedDeclarations API example? 

There is an example https://github.com/vuejs/core/blob/rolldown/scripts/build-types.js#L20
`Vue` tries to use `isolatedDeclarations` to generate .d.ts for all packages.

File: src/blog/2024-09-29-transformer-alpha.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+[@lukeed](https://github.com/lukeed) and [@maraisr](https://github.com/maraisr) are utilizing `oxc-transform`
+for their packages
+[`empathic`](https://github.com/lukeed/empathic) and [`dldr`](https://github.com/maraisr/dldr)
+to transform and generate dts [in a single step](https://github.com/lukeed/empathic/blob/b83a360ff55051590dec19aa913cd12da97fa3f8/scripts/build.ts#L45-L52).
+
+The following example demonstrates emitting `.js` and `.d.ts` in a single transformation step:
+
+```javascript
+import { transform } from "oxc-transform";
+const transformed = transform(entry, source, {
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:46]: I missed this first time around. What does "entry" mean?

```suggestion
const transformed = transform(filePath, sourceCode, {
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+2. [TypeScript Isolated Declarations DTS Emit](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-5.html#isolated-declarations) (`tsc --isolatedDeclarations`).
+
+In this alpha stage, we recommend to experiment these features to significantly speed up continuous integration time.
+
+[Our Benchmark](https://github.com/oxc-project/bench-transformer) shows that:
+
+- For transform, oxc is 4x faster than swc, uses 20% less memory, and is 35 MB smaller in package size (from swc's 37MB).
+- For transform, oxc is 40x faster than babel, uses 70% less memory, and is 19 MB smaller with 168 npm packages less to install.
+- For react development + refresh, oxc is 6x faster than swc and 70x faster than Babel.
+- For tsc's isolated declarations dts emit, oxc is 45x faster on ordinary files, and 20x faster on larger files.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:26]: Make capitalization consistent and clarify comparisons (well clearer in my opinion anyway):

```suggestion
- Transform: Oxc is 4x faster than SWC, uses 20% less memory, and has far smaller package size (2 MB vs SWC's 37 MB).
- Transform: Oxc is 40x faster than Babel, uses 70% less memory, and is 19 MB smaller, with only 2 npm packages to install vs Babel's 170.
- React development + React Refresh: Oxc is 6x faster than SWC, and 70x faster than Babel.
- TS isolated declarations `.d.ts` emit: Oxc is 45x faster than TSC on typical files, 20x faster on larger files.
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+## Usage Examples
+
+### [`oxc-transform`](https://www.npmjs.com/package/oxc-transform) npm package
+
+vue.js is [experimenting](https://github.com/vuejs/core/blob/0895b2624b707ea1e75c41f2e1f75388e7a6f101/scripts/build-types.js#L20) with isolated declarations in their build pipeline.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:32]: Consistent capitalization.

```suggestion
Vue.js is [experimenting](https://github.com/vuejs/core/blob/0895b2624b707ea1e75c41f2e1f75388e7a6f101/scripts/build-types.js#L20) with isolated declarations in their build pipeline.
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+await fs.writeFile("out.d.ts", transformed.declaration);
+```
+
+### [`unplugin-isolated-decl`](https://github.com/unplugin/unplugin-isolated-decl)
+
+`vue-macros` [uses](https://github.com/vue-macros/vue-macros/blob/4247c7ba9189c630111e058245ce1412c8da9229/tsup.config.ts#L10) `unplugin-isolated-decl` as the integration tool for its esbuild plugin.
+
+[@sxzz](https://github.com/sxzz) [reports](https://x.com/sanxiaozhizi/status/1821320327231893600) their `.d.ts` generation time is reduced from 76s to 16s.
+
+### Bazel Build at [Airtable](https://github.com/Airtable)
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:64]:
```suggestion
### Bazel build at [Airtable](https://github.com/Airtable)
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+[@michaelm](https://github.com/MichaelMitchell-at) from [Airtable](https://github.com/Airtable)
+is integrating oxc's isolated declarations dts emit in their CI pipeline within their [bazel build](https://github.com/aspect-build/rules_ts/pull/697).
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:67]:
```suggestion
is integrating Oxc's isolated declarations `.d.ts` emit in their CI pipeline within their [Bazel build](https://github.com/aspect-build/rules_ts/pull/697).
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+```javascript
+import { isolatedDeclaration } from "oxc-transform";
+const dts = isolatedDeclaration(filename, ts);
+```
+
+[@lukeed](https://github.com/lukeed) and [@maraisr](https://github.com/maraisr) are utilizing `oxc-transform`
+for their packages
+[`empathic`](https://github.com/lukeed/empathic) and [`dldr`](https://github.com/maraisr/dldr)
+to transform and generate dts [in a single step](https://github.com/lukeed/empathic/blob/b83a360ff55051590dec19aa913cd12da97fa3f8/scripts/build.ts#L45-L52).
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:42]:
```suggestion
to transform and generate `.d.ts` [in a single step](https://github.com/lukeed/empathic/blob/b83a360ff55051590dec19aa913cd12da97fa3f8/scripts/build.ts#L45-L52).
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+[@sxzz](https://github.com/sxzz) [reports](https://x.com/sanxiaozhizi/status/1821320327231893600) their `.d.ts` generation time is reduced from 76s to 16s.
+
+### Bazel Build at [Airtable](https://github.com/Airtable)
+
+[@michaelm](https://github.com/MichaelMitchell-at) from [Airtable](https://github.com/Airtable)
+is integrating oxc's isolated declarations dts emit in their CI pipeline within their [bazel build](https://github.com/aspect-build/rules_ts/pull/697).
+
+### Rust [`oxc_transformer`](https://crates.io/crates/oxc_transformer) crate
+
+The [Rolldown](https://rolldown.rs) bundler uses the `oxc_transformer` crate [directly](https://github.com/rolldown/rolldown/blob/3213e8fdb9e25f29295b6ec0d92fcc2ce03ce396/crates/rolldown/src/utils/pre_process_ecma_ast.rs#L67-L75).
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:71]:
```suggestion
[Rolldown](https://rolldown.rs) bundler uses the `oxc_transformer` crate [directly](https://github.com/rolldown/rolldown/blob/3213e8fdb9e25f29295b6ec0d92fcc2ce03ce396/crates/rolldown/src/utils/pre_process_ecma_ast.rs#L67-L75).
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+## [Benchmark Results](https://github.com/oxc-project/bench-transformer)
+
+On [GitHub Actions ubuntu-latest](https://github.com/oxc-project/bench-transformer/actions/workflows/ci.yml):
+
+Using the following files:
+
+- [cal.com - UserSettings.tsx](https://github.com/calcom/cal.com/blob/20729b3a4e62c52f49419d2c3b30225f0c7a5936/apps/web/components/getting-started/steps-views/UserSettings.tsx) - a typical 124 lines of tsx code.
+- [TypeScript - parser.ts](https://github.com/microsoft/TypeScript/blob/3ad0f752482f5e846dc35a69572ccb43311826c0/src/compiler/parser.ts) - an atypical large file with 10777 lines.
+- [vuejs - renderer.ts](https://github.com/vuejs/core/blob/cb34b28a4a9bf868be4785b001c526163eda342e/packages/runtime-core/src/renderer.ts) - a large library file with 2550 lines.
+- [AFFiNE - table.tsx](https://github.com/toeverything/AFFiNE/blob/a9b29d24f1f6e5563e43a11b5cbcfb30c9981d25/packages/frontend/core/src/components/affine/page-properties/table.tsx) - a tsx file with 1118 lines.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:82]:
```suggestion
- [cal.com - UserSettings.tsx](https://github.com/calcom/cal.com/blob/20729b3a4e62c52f49419d2c3b30225f0c7a5936/apps/web/components/getting-started/steps-views/UserSettings.tsx) - typical TSX file with 124 lines of code.
- [TypeScript - parser.ts](https://github.com/microsoft/TypeScript/blob/3ad0f752482f5e846dc35a69572ccb43311826c0/src/compiler/parser.ts) - an atypically large TS file with 10777 lines.
- [Vue.js - renderer.ts](https://github.com/vuejs/core/blob/cb34b28a4a9bf868be4785b001c526163eda342e/packages/runtime-core/src/renderer.ts) - a large library file with 2550 lines.
- [AFFiNE - table.tsx](https://github.com/toeverything/AFFiNE/blob/a9b29d24f1f6e5563e43a11b5cbcfb30c9981d25/packages/frontend/core/src/components/affine/page-properties/table.tsx) - TSX file with 1118 lines.
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+oxc > renderer.ts (1.7 ms)
+  3.96x faster than swc (6.6 ms)
+  30.35x faster than babel (55.2 ms)
+
+oxc > table.tsx (1.0 ms)
+  4.54x faster than swc (5.3 ms)
+  59.54x faster than babel (58.5 ms)
+```
+
+Oxc is 4x faster than swc, and 40x faster than Babel.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:104]:
```suggestion
Oxc is 4x faster than SWC, and 40x faster than Babel.
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  20.42x faster than tsc (96.0 ms)
+
+oxc > renderer.ts (0.7 ms)
+  21.70x faster than tsc (22.2 ms)
+
+oxc > table.tsx (1.7 ms)
+  7.24x faster than tsc (22.0 ms)
+```
+
+Compared to tsc, oxc is 45x faster on small files, and 20x faster on larger files.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:122]:
```suggestion
Compared to `tsc`, Oxc is 45x faster on small files, and 20x faster on larger files.
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+## Package size
+
+For package download size, oxc downloads two packages for a total of 2MB.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:126]:
```suggestion
Oxc downloads only two npm packages, a total of 2MB.
```

File: src/blog/2024-09-29-transformer-alpha.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+| ---------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------- |
+| `@oxc-transform/binding-darwin-arm64`                                                    | [2.0 MB](https://www.npmjs.com/package/@oxc-transform/binding-darwin-arm64)                 |
+| `@swc/core-darwin-arm64`                                                                 | [37.5 MB](https://www.npmjs.com/package/@swc/core-darwin-arm64)                             |
+| `@babel/core` + `@babel/preset-env` + `@babel/preset-react` + `@babel/preset-typescript` | [21 MB and 170 packages](https://www.npmjs.com/package/@oxc-transform/binding-darwin-arm64) |
+
+## Memory Usage
+
+Oxc uses less memory.
+
+On `parser.ts` (10777 lines) by using `/usr/bin/time -alh node`:
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:138]:
```suggestion
Memory usage transforming `parser.ts` (10777 lines) - measured using `/usr/bin/time -alh node`:
```

**Review by overlookmotel — COMMENTED** [review]
Just a bunch of nits, mostly relating to consistent formatting/capitalization of product names.

**Comment by Dunqing** [comment]
It looks like we should mention that we already support React Fast Refresh and `vite-plugin-react-oxc`(needs to confirm) is coming soon.

File: src/blog/2024-09-29-transformer-alpha.md [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+We are excited to announce an alpha release for Oxc transformer (a.k.a transpiler).
+
+This release contains three major features:
+
+1. Transforming TypeScript to ESNext.
+2. Transforming React JSX to ESNext, with built-in React Refresh.
+3. [TypeScript Isolated Declarations DTS Emit](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-5.html#isolated-declarations) (`tsc --isolatedDeclarations`).
+
+In this alpha stage, we recommend to experiment these features to significantly speed up continuous integration time.
```
Comment (overlookmotel) [inline src/blog/2024-09-29-transformer-alpha.md:20]:
```suggestion
In this alpha stage, we recommend to experiment with these features to significantly speed up build times.
```

**Pull request merged**